File: java/leetcode/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseList (ListNode head) {
        ListNode currNode = head;
        ListNode nextNode;
        ListNode prevNode = null;
        if (currNode == null) {
            prevNode = currNode;
        } else {
            while (currNode != null) {
                nextNode = currNode.next;
                currNode.next = prevNode;
                prevNode = currNode;
                currNode = nextNode;
            }
        }
        return prevNode;
    }
}

File: ruby/leetcode/reverse-linked-list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
    curr_node = head
    prev_node = nil
    while curr_node
        next_node = curr_node.next
        curr_node.next = prev_node
        prev_node = curr_node
        curr_node = next_node
    end
    return prev_node || curr_node

end
